File: 0078-subsets/0078-subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
function subset(index, subsetArr, nums, result) {

    if ( index >= nums.length ) {
        result.push([...subsetArr]);
        return;
    }
    
    subsetArr.push(nums[index]);
    subset(index+1, subsetArr, nums, result);

    subsetArr.pop();
    subset(index+1, subsetArr, nums, result);

    return result
}
var subsets = function(nums) {
    let subsetArr = [];
    let result = [];
    return subset(0, subsetArr, nums, result);
};

File: 0078-subsets/0078-subsets.java
class Solution {
    public void generateSubsets(List<List<Integer>> result, List<Integer> subResult, int[] nums, int i) {
        if ( i == nums.length ) {
            result.add(new ArrayList<>(subResult));
            return;
        }
        subResult.add(nums[i]);
        generateSubsets(result, subResult, nums, i+1);
        subResult.remove(subResult.size()-1);
        generateSubsets(result, subResult, nums, i+1);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> subResult = new ArrayList<>();
        // Arrays.sort(nums);
        generateSubsets(result, subResult, nums, 0);
        return result;
    }
}